src/compiler/Error.elm

```diff
@@ -116,6 +121,7 @@ type TypeError
     = UnknownName VarName
     | TypeMismatch Type Type
     | OccursCheckFailed Int Type
+    | ListTypeMismatch Int Type Type
```

Comment:
Sorry if I'm commenting on what you'll want to change later! Dunno which parts are still WIP :)

I think this error has a bit better knowledge of the types than the generic `TypeMismatch`.
This one can say "the nth type was Foo, all before were Bar!"... I'd make the parameters a record
```elm
| ListTypeMismatch
    { position : Int
    , type_ : Type
    , previousTypes : Type
    }
```
to prevent the confusion between which `Type` is which, and edit the message to say something like "This element has type Foo, and all the previous elements had type Bar".